ui/button.js

```diff
@@ -56,355 +39,267 @@ $.widget( "ui.button", {
+	_updateIcon: function( icon ) {
+		if ( !this.icon ) {
+			this.icon = $( "<span>" );
+			this._addClass( this.icon, "ui-button-icon", " ui-icon" );
 
-			checked = this.element.is( ":checked" );
-			if ( checked ) {
-				this.buttonElement.addClass( "ui-state-active" );
+			if ( !this.options.showLabel ) {
+				this._addClass( "ui-button-icon-only" );
+			} else {
+				this._addClass( null, "ui-icon-" + this.options.iconPosition );
 			}
-			this.buttonElement.prop( "aria-pressed", checked );
 		} else {
-			this.buttonElement = this.element;
+			this._removeClass( this.icon, null, this.options.icon );
 		}
-	},
-
-	widget: function() {
-		return this.buttonElement;
+		this._addClass( this.icon, null, icon );
+		this.icon.appendTo( this.element );
+		return this;
 	},
 
 	_destroy: function() {
-		this.element
-			.removeClass( "ui-helper-hidden-accessible" );
-		this.buttonElement
-			.removeClass( baseClasses + " ui-state-active " + typeClasses )
-			.removeAttr( "role" )
-			.removeAttr( "aria-pressed" )
-			.html( this.buttonElement.find(".ui-button-text").html() );
+		this.element.removeAttr( "role" );
 
+		if ( this.icon ) {
+			this.icon.remove();
+		}
 		if ( !this.hasTitle ) {
-			this.buttonElement.removeAttr( "title" );
+			this.element.removeAttr( "title" );
 		}
 	},
 
 	_setOption: function( key, value ) {
+		if ( key === "icon" ) {
+			if ( value !== null ) {
+				this._updateIcon( value );
+			} else {
+				this.icon.remove();
+				this._removeClass( "ui-button-icon", " ui-icon-" + this.options.iconPosition );
+			}
+		}
+
+		// Make sure we can't end up with a button that has no text nor icon
+		if ( key === "showLabel" ) {
+			if ( ( !value && this.options.icon ) || value ) {
+				this._toggleClass( this._classes( "ui-button-icon-only" ), null, !value )
+					._toggleClass( null, this.options.iconPosition, value );
+				this._updateTooltip();
+			} else {
+				value = true;
+			}
+		}
+		if ( key === "iconPosition" && this.options.icon ) {
+			this._addClass( null, value )
+				._removeClass( null, this.options.iconPosition );
+		}
+		if ( key === "label" ) {
+			if ( this.isInput ) {
+				this.element.val( value );
+			} else {
+
+				// If there us an icon append it else nothing then append the value
+				// this avoids removal of the icon when setting label text
+				this.element.html( !!this.icon ? "" : this.icon ).append( value );
+			}
+		}
 		this._super( key, value );
 		if ( key === "disabled" ) {
-			this.widget().toggleClass( "ui-state-disabled", !!value );
-			this.element.prop( "disabled", !!value );
-			if ( value ) {
-				if ( this.type === "checkbox" || this.type === "radio" ) {
-					this.buttonElement.removeClass( "ui-state-focus" );
-				} else {
-					this.buttonElement.removeClass( "ui-state-focus ui-state-active" );
-				}
-			}
-			return;
+			this._toggleClass( null, "ui-state-disabled", value ).element[ 0 ].disabled = value;
+			this.element.blur();
```

Comment:
This should be only called when `value` is truthy, otherwise the element can't have focus anyway.